Add 'filter' to 'knn' search section

jtibshirani opened the pull request:
This PR introduces a 'filter' subsection where filters can be provided to the
kNN search:

```
POST index/_search
{
  "knn": {
    "field": "image_vector",
    "query_vector": [0.3f, 0.1f, ...],
    "k": 10,
    "num_candidates": 100,
    "filter": {
      "term": {
       "file_type": "jpeg"
      }
    }
  }
}
```

Like all other queries, the filters are rewritten on the coordinating nodes as
well as data nodes.

Addresses #87625.

jtibshirani committed:
Add 'filter' to 'knn' search section

This PR introduces a 'filter' subsection where filters can be provided to the
kNN search.

Like all other queries, the filters are rewritten on the coordinating nodes as
well as data nodes.

elasticmachine commented:
Pinging @elastic/clients-team (Team:Clients)

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)

jtibshirani committed:
Fix checkstyle

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java:
```diff
@@ -123,6 +167,7 @@ public boolean equals(Object o) {
             && numCands == that.numCands
             && Objects.equals(field, that.field)
             && Arrays.equals(queryVector, that.queryVector)
+            && Objects.equals(filterQueries, that.filterQueries)
```
Should we add `filterQueries` to `hashCode()` as well?

jtibshirani replied:
Yes, thanks for catching that!

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java:
```diff
@@ -151,6 +204,10 @@ public void writeTo(StreamOutput out) throws IOException {
         out.writeVInt(k);
         out.writeVInt(numCands);
         out.writeFloatArray(queryVector);
+        out.writeVInt(filterQueries.size());
```
Should we write filterQueries the same way we read them:
`out.writeNamedWriteableList(filterQueries);`?

jtibshirani replied:
Good point, not sure why I did this manually.

mayya-sharipova reviewed (approved):
@jtibshirani Thanks, great work! I left a couple of comments.

jtibshirani committed:
Address code review comments

jtibshirani commented:
Thanks @mayya-sharipova for the review. I addressed your comments and also updated the feature branch `knn-search` with the latest commits from `main` (that's why there are force pushes).
